sdk/servicebus/Azure.Messaging.ServiceBus/src/Sender/ServiceBusMessageBatch.cs

```diff
@@ -74,14 +77,42 @@ internal ServiceBusMessageBatch(TransportMessageBatch transportBatch)
         ///
         public bool TryAdd(ServiceBusMessage message)
         {
+            if (_locked)
+            {
+                throw new InvalidOperationException(Resources.MessageBatchIsLocked);
+            }
+
             return InnerBatch.TryAdd(message);
         }
 
         /// <summary>
         ///   Performs the task needed to clean up resources used by the <see cref="ServiceBusMessageBatch" />.
         /// </summary>
         ///
-        public void Dispose() => InnerBatch.Dispose();
+        public void Dispose()
+        {
+            if (_locked)
+            {
+                throw new InvalidOperationException(Resources.MessageBatchIsLocked);
+            }
+
+            InnerBatch.Dispose();
+        }
+
+        /// <summary>
+        ///   Clears the batch, removing all messages and resetting the
+        ///   available size.
+        /// </summary>
+        ///
+        internal void Clear()
```

Comment:
Also curious about this. I didn't see an equivalent method for EH. Is there a reason this would only apply to SB?

Reply:
This is part of the discussion in the Event Hubs channel to introduce the ability for callers to clear and reuse a batch.   It won't be invoked internally other than by `Dispose` but is expected to be put forward for community discussion and architect review for both Service Bus and Event Hubs.   Since we just so happened to be in this area and the change was minimal, I implemented it in anticipation so that we just flip visibility `public` and have the feature (and tests) ready.

The Event Hubs equivalent is also in flight, covered in #11203